### ISSUE TITLE
Error when using URI module to send JSON body with variables - TypeError: must be string or buffer, not dict

### ISSUE THREAD

**Issue by lancehudson** [ansible/ansible-modules-core#265]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.7.2
##### Environment:

centos 7
python 2.7.5
##### Summary:

I am getting a TypeError when attempting to send a JSON body in a URI POST that is using a variable.
##### Steps To Reproduce:

run this playbook - https://gist.github.com/lancehudson/3e7e15b2a11f6cc4b4eb#file-playbook-fails

```
bash-4.2$ ansible-playbook -vvvv -i hosts playbook-fails.yml --connection=local
```
##### Expected Results:

no error, post being sent.
##### Actual Results:

Error: TypeError: must be string or buffer, not dict

https://gist.github.com/lancehudson/3e7e15b2a11f6cc4b4eb#file-error
##### Other:

If I hardcode the response it works: https://gist.github.com/lancehudson/3e7e15b2a11f6cc4b4eb#file-playbook-hardcoded

As soon as I introduce an error it fails with the same error: https://gist.github.com/lancehudson/3e7e15b2a11f6cc4b4eb#file-playbook-fails-2

**Comment by flisky** [comment]
Cross reference:
ansible/ansible#9472

Root analysis:
the varname goes through `utils.safe_eval` in template function from "lib/ansible/utils/template.py", and as it succeeds, which converts our json str **back** into a dict.

Dirty hack:
any way to make `safe_eval` fail, such as adding a key with value 'null'.

So can we fix it?

**Comment by dkerwin** [comment]
:+1: Fix would be highly appreciated. Impossible to use variables in uri's json body

**Comment by clickfreak** [comment]
:+1:

**Comment by tomwidmer** [comment]
Does anyone have a workaround for this? Does loading the json from a file using a lookup help? Anything else? Thanks.

**Comment by flisky** [comment]
@tomwidmer, please see my comment "Dirty hack".

**Comment by evgeny-goldin** [comment]
:+1:

**Comment by wybczu** [comment]
:+1:

**Comment by olvesh** [comment]
Imho a better workaround: 

``` yaml
  - name: post stuff in var 'service_body' to uri
    command: >
      curl -XPOST  -d '{{ service_body | to_json }}' http://example.com/rest/service
```

(although it might need some tweaking if you need it to fail on e.g. a 404)

**Comment by smiller171** [comment]
:+1:

This is a real problem for me.

**Comment by smiller171** [comment]
I tried templating the json into a file and templating it into a var, no dice.

**Comment by swordeh** [comment]
+1 on this issue. I'd like to use uri for some CouchDB work, as using command to curl it instead feels a bit pointless if uri could do it.

**Comment by anandnalya** [comment]
:+1:

**Comment by airandfingers** [comment]
+1

**Comment by sbocahu** [comment]
+1

**Comment by sbocahu** [comment]
I propose #1011

**Comment by aioue** [comment]
+1

**Comment by aioue** [comment]
+1 @flisky do you have an example of the hack please?

**Comment by flisky** [comment]
```
---
- hosts: all
  vars:
    body:
      _hack: null  # an extra key with null value, which fails `utils.safe_eval`

  tasks:
    - uri:
        url: "http://httpbin.org/get"
        HEADER_Content-Type: "application/json"
        body: "{{ body | to_json }}"
```

**Comment by aikomastboom** [comment]
patch that works for me is adding the following:

``` python
   if isinstance(body, dict):
       body = json.dumps(body)
```

here: https://github.com/ansible/ansible-modules-core/blob/release1.8.2/network/basics/uri.py#L244

**Comment by abadger** [comment]
I think this was fixed by: https://github.com/ansible/ansible-modules-core/pull/1011  (Will be in the next feature release of ansible... you should be able to grab the referenced uri.py file from the git repo and replace the current module with it if you need to test on 1.9.x though.)  Can someone confirm that the new "body_format" parameter fixes this?

I'll close this in a few days as fixed by that PR if no one reports that it doesn't fix this.

**Comment by smiller171** [comment]
@abadger I haven't tested, but I think this does fix the issue. Unfortunately it has been classified as a new feature instead of as a bugfix, preventing it from going out in the 1.9.1 release.

**Comment by wimnat** [comment]
@abadger I think this is good to be closed now. I tested -https://groups.google.com/forum/m/#!searchin/ansible-devel/Uri$20json/ansible-devel/76oohKlvb-g

**Comment by bcoca** [comment]
closing as per comments

**Comment by saulshanabrook** [comment]
@bcoca Could this be released as a bugfix so that we can use it in the 1.9.x version?

**Comment by arusahni** [comment]
:+1: for a bugfix release.

**Comment by abadger** [comment]
Unfortunately, it is a new feature, not a bugfix.  You can however, download the module from the git repository and drop it into your module path if you want to use it before we make the next feature release.

**Comment by aikomastboom** [comment]
Very interesting how the fix for a bug (found in 1.7.2)  that drops ansible to the prompt gets turned into a feature for the next release (1.9). I expected it to have been marked critical and fixed with high priority at the latest in 1.8.
PS: https://github.com/ansible/ansible-modules-core/pull/1011 does not include any (regression) tests? or is there a seperate git repo holding all the core-module tests?

**Comment by abadger** [comment]
Calling a feature a bugfix is all well and good but it doesn't change the fact that it is a new feature.  Please feel free to utilize my suggestion to put the new module into place in your local module library so that you can get it before the next release.

Tests live in the main repository https://github.com/ansible/ansible .  I don't see that a new test was added when the user submitted their new feature, we were probably so eager to get a PR that closed this issue that we didn't demand a test at that time.  If you'd like to submit a PR for an integration test, the code is here: https://github.com/ansible/ansible/blob/devel/test/integration/roles/test_uri/tasks/main.yml
